### PR TITLE
Google OAuth example

### PR DESCRIPTION
**Pull request by andreyaksenov**
https://github.com/ktorio/ktor-documentation/tree/auth-examples/codeSnippets/snippets/auth-oauth-google

### CODE REVIEW

File: codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    install(Authentication) {
+        oauth("auth-oauth-google") {
+            urlProvider = { "http://localhost:8080/callback" }
+            providerLookup = {
+                OAuthServerSettings.OAuth2ServerSettings(
+                    name = "google",
+                    authorizeUrl = "https://accounts.google.com/o/oauth2/auth",
+                    accessTokenUrl = "https://accounts.google.com/o/oauth2/token",
+                    requestMethod = HttpMethod.Post,
+                    clientId = System.getenv("GOOGLE_CLIENT_ID"),
```
Comment (Stexxe) [inline codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt:26]: We should err If environment variable isn't set.

File: codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                    clientSecret = System.getenv("GOOGLE_CLIENT_SECRET"),
+                    defaultScopes = listOf("https://www.googleapis.com/auth/userinfo.profile")
+                )
+            }
+            client = httpClient
+        }
+    }
+    routing {
+        var principal: OAuthAccessTokenResponse.OAuth2? = null
```
Comment (Stexxe) [inline codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt:35]: That's not a right way to do it. You should save an access token in a user's session.

File: codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        authenticate("auth-oauth-google") {
+            get("/callback") {
+                principal = call.authentication.principal()
+                call.respondRedirect("/hello")
+            }
+        }
+        get("/") {
+            call.respondRedirect("/callback")
```
Comment (Stexxe) [inline codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt:43]: Why do you need this redirect?

File: codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            call.respondText("Hello, ${userInfo.name}!")
+        }
+    }
+}
+
+data class UserInfo(
+    val id: String,
+    val name: String,
+    val given_name: String,
```
Comment (Stexxe) [inline codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt:59]: Code style

**Review by Stexxe — CHANGES_REQUESTED** [review]

File: codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt (hunk 3 above)
Comment (andreyaksenov) [reply codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt:43]: to quickly open the Google authorization page when a user runs an application and clicks `http://0.0.0.0:8080`

File: codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt (hunk 2 above)
Comment (andreyaksenov) [reply codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt:35]: ok, saved a token to cookies

File: codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt (hunk 1 above)
Comment (andreyaksenov) [reply codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt:26]: I agree that such situations should be handled but I'd not add additional checks here to simplify a code snippet in docs. In addition, an application raises a more or less obvious exception in this case (`System.getenv("GOOGLE_CLIENT_ID") must not be null`).

File: codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                    append(HttpHeaders.Authorization, "Bearer ${userSession?.token}")
+                }
+            }
+            call.respondText("Hello, ${userInfo.name}!")
+        }
+    }
+}
+
+data class UserSession(val token: String)
+data class UserInfo(val id: String, val name: String, val given_name: String, val family_name: String, val picture: String, val locale: String)
```
Comment (Stexxe) [inline codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt:62]: The indentation was good, I meant `given_name` and `family_name` identifiers :)

File: codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt (hunk 3 above)
Comment (Stexxe) [reply codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt:47]: It's kinda confusing for a reader that you use `/callback` route to authenticate and as a callback URL. I'd recommend having a separate `/login` route in the `authenticate` block and redirect a user there.

File: codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                val principal: OAuthAccessTokenResponse.OAuth2? = call.authentication.principal()
+                call.sessions.set(UserSession(principal?.accessToken.toString()))
+                call.respondRedirect("/hello")
+            }
+        }
+        get("/") {
+            call.respondRedirect("/callback")
+        }
+        get("/hello") {
+            val userSession: UserSession? = call.sessions.get<UserSession>()
```
Comment (Stexxe) [inline codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt:50]: What will happen if a user doesn't have a valid session with a token?

**Review by Stexxe — CHANGES_REQUESTED** [review]

**Review by Stexxe — APPROVED** [review]

**Pull request merged**